enhancements/olm/boxcutter-library.md

```diff
@@ -0,0 +1,850 @@
+
+```mermaid
+sequenceDiagram
+    participant Admin as Cluster Administrator
+    participant CE as ClusterExtension Controller
+    participant SRG as SimpleRevisionGenerator
+    participant API as Kubernetes API Server
+    participant CER as ClusterExtensionRevision Controller
+    participant RE as RevisionEngine
+    participant R as Managed Resources
+
+    Admin->>API: Create ClusterExtension
+    API-->>CE: Watch event
+    CE->>CE: Resolve bundle via catalogd
+    CE->>CE: Unpack bundle content
+    CE->>SRG: Generate ClusterExtensionRevision
+    SRG->>SRG: Organize manifests into phases
+    SRG-->>CE: Return CER spec
+    CE->>API: Create ClusterExtensionRevision
+    CE->>API: Update CE status (Progressing=True)
+    API-->>CER: Watch event (new CER)
+    CER->>CER: Establish dynamic watches for resource GVKs
+    CER->>RE: Reconcile(revision)
+    RE->>API: Apply Phase 1 objects (CRDs, RBAC)
+    RE->>RE: Evaluate Phase 1 probes
+    RE->>API: Apply Phase 2 objects (Deployments)
+    RE->>RE: Evaluate Phase 2 probes
+    R-->>RE: Probe results (replicas ready)
+    RE-->>CER: Rollout complete
+    CER->>API: Update CER status (Available=True, Succeeded=True)
+    API-->>CE: Watch event (CER status updated)
+    CE->>API: Update CE status (healthy)
+```
+
+#### Actors
+
+- **Cluster Administrator**: Creates/updates `ClusterExtension` resources to
+  install or upgrade operators
+- **ClusterExtension Controller**: Resolves bundles and generates
+  `ClusterExtensionRevision` objects
+- **ClusterExtensionRevision Controller**: Reconciles managed resources to
+  match revision spec
+- **Boxcutter RevisionEngine**: Core library implementing rollout logic and
+  probe evaluation
+- **Kubernetes API Server**: Stores all resources and processes Server-Side
+  Apply patches
+
+#### Workflow: Initial Installation
+
+1. Cluster Administrator creates a `ClusterExtension` resource referencing an
+   operator bundle
+2. `ClusterExtension` controller resolves the bundle reference via catalogd and
+   unpacks content
+3. `ClusterExtension` controller invokes `BoxcutterStorageMigrator` to check for
+   existing Helm Release Secret (none found for new install)
+4. `ClusterExtension` controller uses `SimpleRevisionGenerator` to create
+   `ClusterExtensionRevision` named `<extension-name>-1` with revision=1
+5. `SimpleRevisionGenerator` extracts manifests from bundle, organizes them into
+   phases using PhaseSort (CRDs/RBAC first, workloads later), and sets
+   objectLabels
+6. `ClusterExtension` controller applies `ClusterExtensionRevision` with
+   Server-Side Apply and reports Progressing=True
+7. `ClusterExtensionRevision` controller detects new revision and establishes
+   dynamic watches for all resource GVKs in the revision
+8. `ClusterExtensionRevision` controller invokes RevisionEngine.Reconcile() which
+   iterates through phases sequentially
+9. For each object in Phase 1 (e.g., CRDs, RBAC), RevisionEngine applies via
```

Comment:
This again may be overly implementation-detail-ey but Phase 1 is `Namespaces`, 2 is 'policies' like `NetworkPolicy`, 3 is RBAC, then 4 is CRDs, etc. Not sure if that detail really matters for the purposes of this doc though and would also make the sequence diagram messier 😆